File: solutions/haskell/diamond/1/src/Diamond.hs
module Diamond (diamond) where

diamond :: Char -> Maybe [String]
diamond c
  | c < 'A' || c > 'Z' = Nothing
  | otherwise =
    let delta = (fromEnum c - fromEnum 'A') + 1
        lst =
          map (line delta) [0 .. delta -1]
            ++ map (line delta) [delta -2, delta -3 .. 0]
     in Just lst
  where
    line s n =
      let ch = [toEnum (fromEnum 'A' + n)]
          half = replicate (s - n - 1) ' ' ++ ch ++ replicate n ' '
       in half ++ tail (reverse half)

File: solutions/ruby/diamond/1/diamond.rb
# frozen_string_literal: true

module Diamond
  def self.make_diamond(letter)
    delta = letter.ord - 'A'.ord + 1
    top = (0..delta-1).map { |d| line(delta, d) }
    bottom = (delta-2).downto(0).map { |d| line(delta, d) }
    (top + bottom).join("\n") + "\n"
  end

  def self.line(w, n)
    spc_before = ' ' * (w-n-1)
    ch = ('A'.ord + n).chr
    spc_after = ' ' * n

    half = spc_before + ch + spc_after
    half + half.chop.reverse
  end
end
